golang/cosmos/x/vpurse/module.go

```diff
@@ -97,6 +97,70 @@ func (am AppModule) BeginBlock(ctx sdk.Context, req abci.RequestBeginBlock) {
 
 // EndBlock implements the AppModule interface
 func (am AppModule) EndBlock(ctx sdk.Context, req abci.RequestEndBlock) []abci.ValidatorUpdate {
+	events := ctx.EventManager().GetABCIEventHistory()
+	addressToBalance := make(map[string]sdk.Coins, len(events)*2)
+
+	ensureBalanceIsPresent := func(address string) error {
+		if _, ok := addressToBalance[address]; ok {
+			return nil
+		}
+		account, err := sdk.AccAddressFromBech32(address)
+		if err != nil {
+			return err
+		}
+		coins := am.keeper.GetAllBalances(ctx, account)
+		addressToBalance[address] = coins
+		return nil
+	}
+
+	/* Scan for all the events matching (taken from cosmos-sdk/x/bank/spec/04_events.md):
+	| Type     | Attribute Key | Attribute Value    |
+	| -------- | ------------- | ------------------ |
+	| transfer | recipient     | {recipientAddress} |
+	| transfer | sender        | {senderAddress}    |
+	| transfer | amount        | {amount}           |
+	| message  | module        | bank               |
+	| message  | action        | send               |
+	| message  | sender        | {senderAddress}    |
+
+	### MsgMultiSend
+
+	| Type     | Attribute Key | Attribute Value    |
+	| -------- | ------------- | ------------------ |
+	| transfer | recipient     | {recipientAddress} |
+	| transfer | sender        | {senderAddress}    |
+	| transfer | amount        | {amount}           |
+	| message  | module        | bank               |
+	| message  | action        | multisend          |
+	| message  | sender        | {senderAddress}    |
+	*/
+	// fmt.Printf("have vpurse events: %+v\n", events)
+	for _, event := range events {
+		switch event.Type {
+		case "transfer":
+			for _, attr := range event.Attributes {
```

Comment:
Consider `event.GetAttributes()` instead, to more uniformly use accessor functions, but it's entirely a matter of preference.